contrib/platform/src/com/sun/jna/platform/mac/SystemB.java

```diff
@@ -596,17 +674,28 @@ public IFmsgHdr2(Pointer p) {
     @Deprecated
     int mach_task_self();
 
+    /**
+     * Decrement the target port right's user reference count.
+     *
+     * @param port
+     *            The port holding the right.
+     * @param name
+     *            The port's name for the right.
+     * @return 0 if successful, a {@code kern_return_t} code otherwise.
+     */
+    int mach_port_deallocate(MachPort port, MachPortName name);
```

Comment:
Is there a set of headers, that show a complete set of definitions for `mach_port_t`, `mach_port_name_t`  and `mach_port_deallocate`? gnu mach has a slightly different definition and it much better fits the logic I see:

https://www.gnu.org/software/hurd/gnumach-doc/Port-Destruction.html

```C
kern_return_t mach_port_deallocate (ipc_space_t task, mach_port_t name)
```

The cast above looks very suspicious and I doubt, that it is the right approach. This feels off and I currently don't trust this logic.

I now ran three times around the block and still don't see different definitions.

I think my problem is, that I find mixed results for kernel and user-space.

I would try to ask a C compiler for the sizes of the types (mach_port_t, mach_port_name_t and the other types) and do this once for 32 and once for 64bit. That way at least some more evidence can be retrieved, which definition is correct. I did this once for the base types:

[test.zip](https://github.com/java-native-access/jna/files/3615870/test.zip)

and got interesting results.

Reply:
Well, I had the brilliant idea of just searching my machine's header files instead of the mishmash of google searches.  It looks to me like everything's 32-bit, which is where I think we started this exercise.  

In `port.h`:

```
typedef natural_t mach_port_name_t;
#include <sys/_types.h>
#include <sys/_types/_mach_port_t.h>
```

In `_mach_port_t.h`, with the comment "this definition of mach_port-t is only for user-space"

```
typedef __darwin_mach_port_t mach_port_t;
#include <sys/_types.h>
```

In `sys/_types.h`, with the comment "takes common type definitions that must be used in multiple header files due to XSI, removes them from the system space, and puts them in the implementation space"

```
#include <machine/_types.h>
typedef __darwin_natural_t __darwin_mach_port_name_t;
typedef __dawrin_mach_port_name_t __darwin_mach_port_t;
```

In `machine/_types.h` (for both 32- and 64-bit):

`#include "i386/_types.h"`

In `i386/_types.h`:

`typedef unsigned int __darwin_natural_t;`

In `mach_port.h`:

```
#include <mach/mach_types.h>
#include <mach/port.h>

kern_return_t mach_port_deallocate
(
    ipc_space_t task,
    mach_port_name_t name
);
```

In `mach/mach_types.h`:

```
/*
 * If we are not in the kernel, then these will all be represented by
 * ports at user-space.
 */
typedef mach_port_t             task_t;
typedef mach_port_t             ipc_space_t;
typedef mach_port_t             host_t;
```

Reply:
```
<danielwiddis> [09-16 09:02] ~/test $ cat mach.c
#include <stddef.h>
#include <stdio.h>
#include "test.h"
#include <mach/mach_types.h>
#include <mach/port.h>

int main() {
    mach_port_t m;
    mach_port_name_t n;
    max_align_t t;

    printf("                        sizeof   align\n");
    printf("mach_port_t             %6zu   %5zu\n", sizeof (m), ALIGN(mach_port_t));
    printf("mach_port_name_t        %6zu   %5zu\n", sizeof (n), ALIGN(mach_port_name_t));
    printf("The size of max_align_t %6zu   %5zu\n", sizeof (t), ALIGN(max_align_t));
}

<danielwiddis> [09-16 09:03] ~/test $ gcc -o mach mach.c

<danielwiddis> [09-16 09:03] ~/test $ ./mach
                        sizeof   align
mach_port_t                  4       4
mach_port_name_t             4       4
The size of max_align_t     16      16
```

Reply:
Interesting quotes from [Mach IPC: The Mac OS X Implementation](https://flylib.com/books/en/3.126.1.105/1/) (emphasis mine):

> The integer type used to represent a port name is historically the native integer type for the machine. This type is called `natural_t` and is accessed by including <mach/machine/vm_types.h>, which in turn accesses it from <mach/ppc/vm_types.h> or <mach/i386/vm_types.h> on the PowerPC and x86 versions, respectively, of Mac OS X. With the introduction of the 64-bit Darwin ABI, several Mach data types (such as `vm_offset_t` and `vm_size_t`) have been scaled to be the same size as a pointer. **However, `natural_t` is 32 bits in size regardless of the ABI.**

and

> 9.3.2.1. What's in a Port's Name?
> 
> It is important to realize the different between `mach_port_t` and `mach_port_name_t`: **The two are treated the same in user space but not in the kernel.** A port's name is relevant only in a particular namespace, corresponding to a task. A `mach_port_name_t` represents the local, namespace-specific identity of a port, without implying any associated rights. A `mach_port_t` represents a reference added or deleted to a port right. Such a reference is represented in user space by returning the name of the right (or many rights) that was altered within the task's IPC space, **which is why it is the same as a `mach_port_name_t` in user space.**

Reply:
Thank you for digging through this. This indeed points to (only user-space):
```
 sizeof(mach_port_t) == sizeof(mach_port_name_t) == sizeof(int) == 4
```
Assuming the output of the C code run was done with a 64 bit target, that would prove that both types are not based on pointers. Great :-)

Reply:
Yep:

> <danielwiddis> [09-16 12:43] ~/test $ gcc -v
> Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
> Apple LLVM version 10.0.1 (clang-1001.0.46.4)
> **Target: x86_64-apple-darwin18.7.0**

And the compiler yells at me when I try to use `i386` that it's deprecated (but I get the same results).

So it looks like I can just go back and put in `int` for all the ports (and no need to remap the existing `mach_xxxx_port()` return values).